design/auto-binding.md

```diff
@@ -0,0 +1,199 @@
+
+### Web app and database
+
+Let's use web app and database use case as an example.
+
+We have a db component like below.
+
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: Component
+metadata:
+  name: db
+spec:
+  workload:
+    apiVersion: database.alibaba.crossplane.io/v1alpha1
+    kind: RDSInstance
+    metadata:
+      name: rdspostgresql
+    spec:
+      forProvider:
+        engine: PostgreSQL
+        engineVersion: "9.4"
+        dbInstanceClass: rds.pg.s1.small
+        dbInstanceStorageInGB: 20
+        securityIPList: "0.0.0.0/0"
+        masterUsername: "test123"
+      writeConnectionSecretToRef:
+        namespace: default
+        name: alibaba-rdspostgresql-conn
+      providerRef:
+        name: alibaba-provider
+      reclaimPolicy: Delete
+```
+
+And a web component which will use secret from db.
+
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: Component
+metadata:
+  name: webapp
+spec:
+  workload:
+    apiVersion: serving.knative.dev/v1
+    kind: Configuration
+    metadata:
+      name: webapp
+      namespace: default
+    spec:
+      template:
+        spec:
+          containers:
+            - image: oamdev/postgresql-flask-web-application:v0.1
+              name: webapp
+              env:
+                - name: DB_HOST
+                  valueFrom:
+                    secretKeyRef:
+                      key: endpoint
+                - name: DB_USER
+                  valueFrom:
+                    secretKeyRef:
+                      key: username
+                - name: DB_PASSWORD
+                  valueFrom:
+                    secretKeyRef:
+                      key: password
+              ports:
+                - containerPort: 80
+                  name: http1 # Must be one of "http1" or "h2c" (if supported). Defaults to "http1".
+          timeoutSeconds: 600
+```
+
+And with an ApplicationConfiguration:
+
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: ApplicationConfiguration
+metadata:
+  name: knative-postgresql-appconfig
+spec:
+  components:
+    - componentName: webapp
+      dataInputs:
+        - valueFrom:
+            dataOutputName: alibaba-rdspostgresql-conn
+          toFieldPaths:
+            - spec.containers[0].env[0].valueFrom.secretKeyRef.name
+            - spec.containers[0].env[1].valueFrom.secretKeyRef.name
+            - spec.containers[0].env[2].valueFrom.secretKeyRef.name
+
+    - componentName: db
+      dataOutputs:
+        - name: alibaba-rdspostgresql-conn
+          fieldPath: spec.writeConnectionSecretToRef.name
+```
+
+In this case, Developer of web component don't need to specify name of secret, but he has to specify the key of secret.
+
+But if the developer don't know the secret name, how can he know the key? So I propose we should automatically bind
+secret name along with the key.
+
+## Proposal
+
+The overall idea is to add more binding ways into `dataInputs` field.
+
+For k8s `secret`, we can add `toSecretKeyRef` field into `dataInputs`:
```

Comment:
At the first glance, I think it is for workload `db` and find it doesn't make sense. And then I find it should be appended for workload `webapp`. So just for clarification, I made the modification:)

Reply:
good